juspay/hyperswitch
# feat(connector): [BANKOFAMERICA] Implement 3DS flow for cards

## Type of Change



## Description

Implement 3DS flow for cards

```
Legend: 
UB: User Browser
M: Merchant
HS: Hyperswitch
CS: Bankofamerica/Cybersource Connector
```

![3DS FLOW DIAGRAM](https://github.com/juspay/hyperswitch/assets/41580413/9f2e0e71-65df-4310-84cf-1419090a680f)


### Additional Changes





## Motivation and Context

https://github.com/juspay/hyperswitch/issues/3342

## How did you test it?


### QA Testing

![Screenshot 2024-01-12 at 5 35 04 PM](https://github.com/juspay/hyperswitch/assets/41580413/bdab526f-736b-44c1-8904-3ae51e9781f6)
![Screenshot 2024-01-12 at 5 33 54 PM](https://github.com/juspay/hyperswitch/assets/41580413/55ca3a43-b8de-43f8-b463-3555a8e79e21)
![Screenshot 2024-01-12 at 5 33 10 PM](https://github.com/juspay/hyperswitch/assets/41580413/6a9d3d59-1a7d-411d-b5e3-61f59376e9c8)

![Screenshot 2024-01-12 at 5 33 28 PM](https://github.com/juspay/hyperswitch/assets/41580413/3ea39f9f-2ea1-4d25-a85e-fcb9b9522e31)

- Test Non-3DS flows to confirm they are not affected
- Test 3DS transactions

All the possible 3DS test cases and test cards are available here: https://developer.cybersource.com/content/dam/docs/cybs/en-us/payer-authentication/developer/all/rest/payer-auth.pdf

3DS Payment Curl:
`curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data-raw '{
  "amount": 1404,
  "currency": "USD",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "customer_id": "CustomerX",
  "email": "guest@example.com",
  "name": "John Doe",
  "phone": "999999999",
  "phone_country_code": "+1",
  "description": "Its my first payment request",
  "authentication_type": "three_ds",
  "return_url": "https://google.com",
  "payment_method": "card",
  "payment_method_type": "credit",
  "payment_method_data": {
    "card": {
       "card_number": "340000000001098",    
      "card_exp_month": "12",
      "card_exp_year": "30",
      "card_holder_name": "joseph",
      "card_cvc": "123"
    }
  },
  "billing": {
    "address": {
      "line1": "sdv",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "46282",
      "country": "US",
      "first_name": "joseph",
      "last_name": "ewcjwd"
    },
    "phone": {
      "number": "8056594427",
      "country_code": "+97"
    }
  },
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "metadata": {
    "count_tickets": "1",
    "transaction_number": 2233
  },
  "business_label": "food",
  "business_country": "US"
}'`

## Checklist
